query-service-impl/src/main/java/org/hypertrace/core/query/service/postgres/PostgresBasedRequestHandler.java

```diff
@@ -0,0 +1,634 @@
+  private static final String COUNT_COLUMN_NAME_CONFIG_KEY = "countColumnName";
+  private static final String START_TIME_ATTRIBUTE_NAME_CONFIG_KEY = "startTimeAttributeName";
+  private static final String SLOW_QUERY_THRESHOLD_MS_CONFIG = "slowQueryThresholdMs";
+
+  private static final int DEFAULT_SLOW_QUERY_THRESHOLD_MS = 3000;
+  private static final Set<Operator> GTE_OPERATORS = Set.of(Operator.GE, Operator.GT, Operator.EQ);
+
+  private final String name;
+  private TableDefinition tableDefinition;
+  private Optional<String> startTimeAttributeName;
+  private QueryRequestToPostgresSQLConverter request2PostgresSqlConverter;
+  // The implementations of ResultSet are package private and hence there's no way to determine the
+  // shape of the results
+  // other than to do string comparison on the simple class names. In order to be able to unit test
+  // the logic for
+  // parsing the Postgres response we need to be able to mock out the ResultSet interface and hence
+  // we
+  // create an interface
+  // for the logic to determine the handling function based in the ResultSet class name. See usages
+  // of resultSetTypePredicateProvider
+  // to see how it used.
+  private final ResultSetTypePredicateProvider resultSetTypePredicateProvider;
+  private final PostgresClientFactory postgresClientFactory;
+
+  private final JsonFormat.Printer protoJsonPrinter =
+      JsonFormat.printer().omittingInsignificantWhitespace();
+
+  private Timer postgresQueryExecutionTimer;
+  private int slowQueryThreshold = DEFAULT_SLOW_QUERY_THRESHOLD_MS;
+
+  PostgresBasedRequestHandler(String name, Config config) {
+    this(name, config, new DefaultResultSetTypePredicateProvider(), PostgresClientFactory.get());
+  }
+
+  PostgresBasedRequestHandler(
+      String name,
+      Config config,
+      ResultSetTypePredicateProvider resultSetTypePredicateProvider,
+      PostgresClientFactory postgresClientFactory) {
+    this.name = name;
+    this.resultSetTypePredicateProvider = resultSetTypePredicateProvider;
+    this.postgresClientFactory = postgresClientFactory;
+    this.processConfig(config);
+  }
+
+  private void initMetrics() {
+    // Registry the latency metric with handler as a tag.
+    this.postgresQueryExecutionTimer =
+        PlatformMetricsRegistry.registerTimer(
+            "postgres.query.latency", Map.of("handler", name), true);
+  }
+
+  @Override
+  public String getName() {
+    return name;
+  }
+
+  @Override
+  public Optional<String> getTimeFilterColumn() {
+    return this.startTimeAttributeName;
+  }
+
+  private void processConfig(Config config) {
+
+    if (!config.hasPath(TENANT_COLUMN_NAME_CONFIG_KEY)) {
+      throw new RuntimeException(
+          TENANT_COLUMN_NAME_CONFIG_KEY + " is not defined in the " + name + " request handler.");
+    }
+
+    String tenantColumnName = config.getString(TENANT_COLUMN_NAME_CONFIG_KEY);
+    String countColumnName =
```

Comment:
want to consider `countColumnName` as Optional as well?

Reply:
Done